Tests/29 add unit tests

insightsurge opened the pull request:
*Have you read the [Contributing Guidelines](https://github.com/Appsilon/.github/blob/main/CONTRIBUTING.md)?* Yes

Closes #29 

## Changes description
Expanded unit test coverage for the package:
- **EmptyStateManager R6 Class:** 
   - Tests if the class has been instantiated
   - Tests if public members exists
   - Tests if private members exist
   - Tests if it can be cloned and cloned obj passes above tests.
   - Tests if the class is locked
- **use_empty_state():** There was not much I can do as the function takes no params (any suggestions are welcome)
   - Tests if source files (css and js) are added properly
   - Tests if dependencies are added properly

insightsurge committed:
Tests added

- Added tests for EmptyStateManager R6 Class
- Added tests for use_empty_state()

insightsurge committed:
lint fix

insightsurge committed:
test coverage workflow removede

insightsurge committed:
Revert "test coverage workflow removede"

This reverts commit 9438b3610e83038db96d707255d14961bb14c215.

insightsurge committed:
added dependencies

insightsurge committed:
syntax erros fix

jakubnowicki committed:
Merge branch 'main' into tests/29-add-unit-tests

jakubnowicki reviewed (changes requested):
This is a good start, but for `EmptyStateManager` you need to test the actual logic, not `R6` features. Please check comments for more details.

Also, if you want to use BDD (which is IMO a good approach), please use `describe` - `it` construction (instead of `test_that`) - more details [here](https://testthat.r-lib.org/reference/describe.html).

insightsurge committed:
test cases updated

- removed r6 structure testcases
- test cases are following BDD structure
- additional tests added for show and hide members (using shinytest2)

insightsurge commented on tests/testthat/test-empty_state.R:
```diff
@@ -0,0 +1,86 @@
+    public_member_tests(manager_class = test_class_cloned)
+    private_member_tests(manager_class = test_class_cloned,
+                         id_attr = "test_id")
+  })
```
Make sense, I've removed this test in latest commit

insightsurge commented on tests/testthat/test-empty_state.R:
```diff
@@ -0,0 +1,86 @@
+describe("EmptyStateManager", {
```
I've added additional tests to test the functioning of show and hide members.

insightsurge commented on tests/testthat/test-empty_state.R:
```diff
@@ -0,0 +1,86 @@
+                 as.character(html_attr))
+    expect_equal(manager_class$.__enclos_env__$private$.color,
+                 color_attr)
+  }
```
I wanted to reuse those tests on cloned object too. But now that I've removed that test completely, there's no point in having these functions.

insightsurge committed:
shinytest2 testcases for show and hide, lint fix

insightsurge committed:
reactable removed, now using renderTable from shiny

insightsurge committed:
button not click test added

insightsurge committed:
shinytest2 setup, adding additional tests

- additional test added (div doesn't exist if button is not clicked)
- shinytest2 setup file added
- test app moved to helpers.R
- gitignore file updated accordingly

insightsurge committed:
CRAN note fix attempt

- added setup-disable-crashpad.R file

insightsurge committed:
CRAN note fix attempt 2

insightsurge committed:
CRAN note fix - DESCRIPTION Updated

insightsurge committed:
Revert "CRAN note fix - DESCRIPTION Updated"

This reverts commit 0022754fa97e54ae7715149181d73d328dc09af6.

insightsurge committed:
Revert "CRAN note fix attempt 2"

This reverts commit b2332d2583e8154513d767c438ad9320b6d9b8f8.

insightsurge committed:
Revert "CRAN note fix attempt"

This reverts commit 36c4aafb28ab704f0905aadbbc7668acb19473c2.

insightsurge committed:
Disable crash reporter + remove temp dir after test

jakubnowicki commented on tests/testthat/test-empty_state.R:
```diff
@@ -0,0 +1,85 @@
+
+  it("should contain default_empty_state_component when no content is passed", {
+    test_class <- EmptyStateManager$new("test_id")
+    expect_equal(test_class$.__enclos_env__$private$.html_content,
+                 as.character(default_empty_state_component()))
```
```suggestion
    expect_equal(
      test_class$.__enclos_env__$private$.html_content,
      as.character(default_empty_state_component())
    )
```

jakubnowicki commented on tests/testthat/test-empty_state.R:
```diff
@@ -0,0 +1,85 @@
+  })
+
+  it("should contain passed color", {
+    test_class <- EmptyStateManager$new("test_id",
+                                        color = "navy")
```
```suggestion
    test_class <- EmptyStateManager$new("test_id", color = "navy")
```

jakubnowicki commented on tests/testthat/test-empty_state.R:
```diff
@@ -0,0 +1,85 @@
+  it("should contain passed color", {
+    test_class <- EmptyStateManager$new("test_id",
+                                        color = "navy")
+    expect_equal(test_class$.__enclos_env__$private$.color,
+                 "navy")
```
```suggestion
    expect_equal(test_class$.__enclos_env__$private$.color, "navy")
```

jakubnowicki commented on tests/testthat/test-empty_state.R:
```diff
@@ -0,0 +1,85 @@
+      "<div class=\"empty-state-content\"><div class=\"myDiv\"></div></div>"
+    app <- shinytest2::AppDriver$new(test_app(), name = "test")
+    app$click("show")
+    expect_equal(app$get_html(selector = ".empty-state-content"),
+                 as.character(expected_div))
```
```suggestion
    expect_equal(
      app$get_html(selector = ".empty-state-content"),
      as.character(expected_div)
    )
```

jakubnowicki commented on tests/testthat/test-empty_state.R:
```diff
@@ -0,0 +1,85 @@
+
+describe("use_empty_state()", {
+  test_func <- use_empty_state()
+  src_files <- list("emptystate.css",
+                    "emptystate.js")
```
```suggestion
  src_files <- list(
    "emptystate.css",
    "emptystate.js"
  )
```

jakubnowicki commented on tests/testthat/test-empty_state.R:
```diff
@@ -0,0 +1,85 @@
+    expect_error(test_class$show <- function() TRUE)
+  })
+
+  it("checks if show functionality work when button is clicked", {
```
Please modify this description so it is more generic - the test is not about the button, but about the show functionality and if it works when triggered. It doesn't matter, what triggers it.

jakubnowicki commented on tests/testthat/test-empty_state.R:
```diff
@@ -0,0 +1,85 @@
+    app$stop()
+  })
+
+  it("checks if show functionality work when button is not clicked", {
+    expected_div <-
+      "<div class=\"empty-state-content\"><div class=\"myDiv\"></div></div>"
+    app <- shinytest2::AppDriver$new(test_app(), name = "test")
+    testthat::expect_false(
+      identical(app$get_html(selector = ".empty-state-content"),
+                as.character(expected_div)))
+    app$stop()
+  })
```
1) Same as for the "show" test - the description should state that the empty state component is hidden if not triggered.
2) This can be simplified - you don't need to specify the exact component, as it should be just NULL. You can use the same test as in the "hide" test.

jakubnowicki commented on tests/testthat/setup-disable-crashpad.R:
Please add a comment that describes the source of this file (as it was copied from `shinytest2`).

jakubnowicki commented on tests/testthat/test-empty_state.R:
```diff
@@ -0,0 +1,85 @@
+    expect_equal(!!as.character(test_dep), !!expect)
+  })
+
+  test_that("should add dependencies properly", {
```
```suggestion
  it("should add dependencies properly", {
```

jakubnowicki reviewed (changes requested):
Overall looks good. Most comments are the suggestions for the code style. Apart from that, only some minor tuning is required.

insightsurge committed:
review changes

- suggestions incorportated
- test description updated
- crashpad disable file source and references updated

jakubnowicki reviewed (changes requested):
The CI passes, but logs show some errors:
https://github.com/Appsilon/shiny.emptystate/actions/runs/5553699568/jobs/10142568414#step:5:159
https://github.com/Appsilon/shiny.emptystate/actions/runs/5553699568/jobs/10142568641#step:5:159
https://github.com/Appsilon/shiny.emptystate/actions/runs/5553699568/jobs/10142568748#step:5:159
https://github.com/Appsilon/shiny.emptystate/actions/runs/5553699568/jobs/10142568882#step:5:154
https://github.com/Appsilon/shiny.emptystate/actions/runs/5553699568/jobs/10142568538#step:5:167

Also, the windows CI is not able to launch the browser (and the test is skipped):
https://github.com/Appsilon/shiny.emptystate/actions/runs/5553699568/jobs/10142568538#step:5:158

insightsurge committed:
removed command of manually closing the browser

insightsurge commented:
Looks like manually disabling browser was causing issue, I've removed it and checks are passing now. Windows test which was skipping is also working fine now.

- **Mac:** https://github.com/Appsilon/shiny.emptystate/actions/runs/5572881630/jobs/10179425814#step:5:158
- **Win:** https://github.com/Appsilon/shiny.emptystate/actions/runs/5572881630/jobs/10179425973#step:5:160
- **Ubuntu devel:** https://github.com/Appsilon/shiny.emptystate/actions/runs/5572881630/jobs/10179426167#step:5:158
- **Ubuntu rel:** https://github.com/Appsilon/shiny.emptystate/actions/runs/5572881630/jobs/10179426281#step:5:158
- **Ubuntu oldrel:** https://github.com/Appsilon/shiny.emptystate/actions/runs/5572881630/jobs/10179426424#step:5:153

jakubnowicki reviewed (approved):
Great!
